xmppserver/src/main/java/org/jivesoftware/openfire/websocket/XmppWebSocket.java

```diff
@@ -125,14 +127,21 @@ public void onTextMethod(String stanza)
     public void onError(Throwable error)
     {
         Log.error("Error detected; session: " + wsSession, error);
-        closeStream(new StreamError(StreamError.Condition.internal_server_error));
-        try {
-            if (wsSession != null) {
-                wsSession.disconnect();
+
+        // Handle asynchronously, to prevent deadlocks. See OF-2473.
+        HttpBindManager.getInstance().getSessionManager().execute(() -> {
+            try {
+                Log.debug("Attempting to close session on which an error occurred.");
+                closeStream(new StreamError(StreamError.Condition.internal_server_error));
```

Comment:
Does closeStream() do the right thing if it's already been closed - now things are no longer synchronous need to be more careful over the processing here

Reply:
I think you're right. By making things more asynchronous, more guards are probably in order. I've added some `synchronous` blocks.

Sideways-related: The code is the exact same snippet that was there before, now only wrapped in a Runnable. Even the old code had some apparent duplication going on (`closeStream` shuts down the session, while that's also done a few lines further down?).